cpp/include/cudf/column/column_device_view.cuh

```diff
@@ -345,6 +369,37 @@ class alignas(16) column_device_view : public detail::column_device_view_base {
   using const_pair_rep_iterator =
     thrust::transform_iterator<detail::pair_rep_accessor<T, has_nulls>, count_it>;
 
+  /**
+   * @brief Return a optional iterator to the first element of the column.
+   *
+   * Dereferencing the returned iterator returns a `thrust::optional<T>`.
+   *
+   * When the element of an iterator contextually converted to bool, the conversion returns true
+   * if the object contains a value and false if it does not contain a value.
+   *
+   * make_optional_iterator supports three different `null` value modes:
+   *
+   * - `HAS_NULLS` means that the column supports nulls and has null values, therefore
+   *    the optional might not contain a value
+   *
+   * - `NO_NULLS` means that the column has no null values, therefore the optional will
+   *    always have a value
+   *
+   * - `RUNTIME`(default) defers the assumption of nullability to runtime with the users stating
+   *    on construction of the iterator if column has nulls.
+   *
+   * @throws cudf::logic_error if the column is not nullable, and `HAS_NULLS` mode used
+   * @throws cudf::logic_error if the column is not nullable, and `RUNTIME` mode used and
+   *         the user has stated nulls exist
+   * @throws cudf::logic_error if column datatype and Element type mismatch.
+   */
+  template <typename T, optional_mode MODE = optional_mode::RUNTIME>
+  const_optional_iterator<T, MODE> optional_begin(bool has_nulls = true) const
```

Comment:
Hm, I think we ideally want when MODE is `NO_NULLS/HAS_NULLS`, then `optional_begin/end` shouldn't have a bool parameter at all. Otherwise you could do something confusing like:
```
col.optional_begin<optional_mode::HAS_NULLS>(false);
```

Reply:
I don't think we can partially specialize template functions, so do you like this API:
```cxx
col.optional_begin(optional_mode::HAS_NULLS{});
col.optional_begin(optional_mode::RUNTIME{}, true);
col.optional_begin(true);  // -> Inferred to be `RUNTIME`
```

Reply:
Yeah, you'd have to do it with SFINAE: https://godbolt.org/z/8h7bzq16f

Reply:
Hm, but I like where your head is at. You're suggesting tag-dispatch instead of template params?

Reply:
Yea I was thinking tag dispatch, but that wont work as shown by https://godbolt.org/z/5z57MzbYj since the overload functions don't have matching parameters.

Looks like your SFINAE approach is the way to go.

Reply:
Your tag-dispatch works, you just forgot to put the tag types in the right place: https://godbolt.org/z/h7KsK4jeY

Reply:
So do you have a preference on the tag dispatch versus SFINAE. I personally prefer tag dispatch.

Reply:
I think the tag-dispatch is definitely cleaner than the SFINAE. I'd go with that.

Reply:
I think I'd take inspiration from `std::span` and `dynamic_extent` and use `DYNAMIC` instead of `RUNTIME`.

Reply:
```
namespace nullable{
  struct YES;
  struct NO;
  struct DYNAMIC;
};
```

The issue is that "nullable" doesn't mean the same thing as "has nulls".

Reply:
@codereport would love your input on the bikeshedding here.

Reply:
I prefer `has_nulls` over `nullable` as that part aligns with the intent of talking about the state of the values of the container, not if the container has support.

Reply:
Yum.

Reply:
Re: naming, I view `RUNTIME` and `DYNAMIC` as interchangeable, and if `std::span` has precedence for `DYNAMIC` I would say go with that. It arguably has precedence with `dynamic_cast` & `dynamic_pointer_cast` as well